common/serial/serial.cpp

```diff
@@ -0,0 +1,418 @@
+//
+// Bareflank Hypervisor
+//
+// Copyright (C) 2015 Assured Information Security, Inc.
+// Author: Rian Quinn        <quinnr@ainfosec.com>
+// Author: Brendan Kerrigan  <kerriganb@ainfosec.com>
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#include <serial.h>
+#include <arch_linux.h>
+
+struct __divisor_rate divisor_table[] = {
+    {50, 0x900},
+    {75, 0x600},
+    {110, 0x417},
+    {150, 0x300},
+    {300, 0x180},
+    {600, 0xC0},
+    {1200, 0x60},
+    {1800, 0x40},
+    {2000, 0x3A},
+    {2400, 0x30},
+    {3600, 0x20},
+    {4800, 0x18},
+    {7200, 0x10},
+    {9600, 0xC},
+    {19200, 0x6},
+    {38400, 0x3},
+    {57600, 0x2},
+    {115200, 0x1},
+    {0, 0},
+};
+
+struct __parity_bits parity_table[] = {
+    { NONE, 0x00},
+    { ODD,  0x08},
+    { EVEN, 0x18},
+    { MARK, 0x28},
+    { SPACE, 0x38},
+    { __PARITY_MAX, 0x00 },
+};
+
+struct __data_bits data_size_table[] = {
+    { 5, 0x00 },
+    { 6, 0x01 },
+    { 7, 0x02 },
+    { 8, 0x03 },
+    { 0, 0x00 },
+};
+
+serial_port::serial_port(uint8_t port, uint32_t baud, uint8_t data_size, PARITY_MODE parity, uint8_t stop_bits) : port_io(new arch_linux)
+{
+    TRACE();
+    switch(port)
+    {
+        case 1:
+        {
+            m_port = COM1_IO_PORT;
+            break;
+        }
+        case 2:
+        {
+            m_port = COM2_IO_PORT;       
+            break;
+        }
+        case 3:
+        {
+            m_port = COM3_IO_PORT;
+            break;
+        }
+        case 4:
+        {
+            m_port = COM4_IO_PORT;
+            break;
+        }
+        default:
+        {
+            // Error case, not a valid port
+            m_port = 0xFFFF;
+            return;
+        }
+    }
+
+    m_baud = baud;
+    m_parity = parity;
+    m_data_size = data_size;
+    m_stop_bits = stop_bits;
+}
+
+serial_port::~serial_port(void)
+{
+
+}
+
+int32_t
```

Comment:
Instead of using int32_t, use an enum that is scoped to the class itself. For example:

```
class serial
{
public:

    enum error
    {
        success = 0,
        failure = 1
    }

public:

    serial() {}
    ~serial() {}

    serial::error do_something() { return success };
}
```